full-width image blocks on blog page

danielfmiranda opened the pull request:
Closes #6676 

**Link to sample test page**:[ Link to Test Blog Page](https://foundation-s-full-width-ct581i.herokuapp.com/en/blog/initial-test-blog-post-with-fixed-title/)

Steps to test:
1.) Click link above to visit a test blog page
2.) Note the full-width image, and how it responds to different screen sizes


**Note**: The wagtail tests are going to fail as I have not added the migrations to source control yet, once approved I will pull the latest migrations from master, run makemigrations, and then merge them in to ensure that they are up to date


## Checklist

**Changes in Models:**
- [x] [Are my changes backward-compatible](https://github.com/mozilla/foundation.mozilla.org/blob/master/docs/workflow.md#django-migrations-what-to-do-when-working-on-backward-incompatible-migrations). If not, did I schedule a deploy with the rest of the team?

danielfmiranda commented on network-api/networkapi/wagtailpages/pagemodels/customblocks/annotated_image_block.py:
```diff
@@ -15,8 +17,27 @@ class AnnotatedImageBlock(ImageBlock):
         icon = 'image'
         template = 'wagtailpages/blocks/annotated_image_block.html'
         help_text = 'Design Guideline: Please crop images to a 16:6 aspect ratio when possible.'
+
```
Function that prevents user from selecting wide and full_width, only will let them select one or the other

danielfmiranda commented on network-api/networkapi/wagtailpages/templates/wagtailpages/campaign_page.html:
```diff
@@ -8,91 +8,95 @@
-    col-12
-  {% endif %}"
->
+<div class="container">
```
since we removed the container and row divs from modular page to enable full-width content on blog pages, we are re-adding them here in campaign _page and some other files as well to ensure they work like before

danielfmiranda commented on source/sass/wagtail/blocks.scss:
```diff
@@ -2,12 +2,20 @@
 }
 
-.wide {
+.full-width {
```
renaming wide to full-width, as this is the class that allows us to have the end to end content

danielfmiranda commented:
Hi @sabrinang! I had a quick question about this PR, if you visit the test page in the description above, you can see that the wide and full-width images are pretty tall. Is this desired, or, should we implement a max height for these? Thank you!

sabrinang commented:
@danielfmiranda Thanks for pointing this out, and agreed it is looking too tall. Is it possible for that the height adapts to the ratio of the image uploaded or will it be cropped to an indicated max height? This would also be useful to document too so we know what dimensions to export/recommend and if there is a focus area to keep in mind.

sabrinang reviewed (changes requested):
wide and full width image height to be adjusted from previous comment

danielfmiranda commented:
Hi @sabrinang I believe at the moment we set the [max height for images to 800px](https://github.com/mozilla/foundation.mozilla.org/blob/baa918ccff6baca6aa6ed0a496580b59da60ceb7/source/sass/wagtail.scss#L75).

Regarding the documentation, I can update the help text underneath the checkbox to mention that full-width images have a max-height of 800.

<img width="588" alt="Screen Shot 2021-06-30 at 6 52 11 PM" src="https://user-images.githubusercontent.com/18314510/124053024-50fd3e00-d9d4-11eb-90e1-38dd6166f066.png">

If this sounds like a good solution, I can make those changes and re-request review 😄 
Thanks!

Pomax commented on network-api/networkapi/wagtailpages/pagemodels/customblocks/annotated_image_block.py:
```diff
@@ -15,8 +17,27 @@ class AnnotatedImageBlock(ImageBlock):
+    def clean(self, value):
+        errors = {}
+
+        if value.get("wide_image") and value.get("full_width_image") is True:
```
We should be able to just say `if value.get("wide_image") and value.get("full_width_image"):` here.

Pomax commented on network-api/networkapi/wagtailpages/templates/wagtailpages/blocks/annotated_image_block.html:
```diff
@@ -1,13 +1,19 @@
 {% block main_block_class %}
-{% if value.wide_image %} wide {% else %} streamfield-content {% endif %}
+streamfield-content 
+  {% if value.full_width_image %} full-width
+  {% elif value.wide_image %} wide
```
let's unindent these

danielfmiranda replied:
Done! Marking as resolved

Pomax commented on network-api/networkapi/wagtailpages/templates/wagtailpages/campaign_page.html:
```diff
@@ -8,91 +8,95 @@
+            {% mini_site_horizontal_nav page %}
+            </div>
+          {% endblock %}
+            {% block html_content %}
+            <div class="cms campaign-page-content wide">
+              {% for block in page.body %}
+                {% if block.block_type == 'paragraph'%}
+                  <div class="container {{block.block_type}}-block">
+                    <div class="row justify-content-center">
+                      <div class="streamfield-content">
+                          {% include_block block%}
+                      </div>
+                    </div>
+                  </div>
+                {% else %}
+                  {% include_block block with parent_page=page %}
+                {% endif %}
+              {% endfor %}
+            </div>
+            {% endblock %}
```
we should unindent the block/endblock here

Pomax reviewed (changes requested):
just a few minor things

danielfmiranda commented:
Hi @Pomax I have implemented your suggestions and have resubmitted for review!

danielfmiranda commented:
Hi @kristinashu! 

For this PR, I have added the following as a fix to the full-width height coversation above:
- Help text on the "full-width image" checkbox to remind the user that the image they upload to be full width should be using the 16:9 aspect ratio. 
<img width="588" alt="Screen Shot 2021-07-08 at 8 51 00 AM" src="https://user-images.githubusercontent.com/18314510/124953208-c87d2f80-dfc9-11eb-8e19-c48e21d1273f.png">

-  a max-height on desktop for full-width images to 730px to avoid very tall images taking up too much page space.

(Desktop with max-height)
![Screenshot 2021-07-08 at 08-53-28 Initial test blog post with fixed title](https://user-images.githubusercontent.com/18314510/124953513-142fd900-dfca-11eb-83ea-cf10519a6aa7.png)

(tablet)
![Screenshot 2021-07-08 at 08-54-53 Initial test blog post with fixed title](https://user-images.githubusercontent.com/18314510/124953673-332e6b00-dfca-11eb-90c2-713c25cba450.png)

(Mobile)
![Screenshot 2021-07-08 at 08-55-23 Initial test blog post with fixed title](https://user-images.githubusercontent.com/18314510/124953756-44777780-dfca-11eb-90a2-5cc53a1a6d23.png)

I am working on getting the review-app back up so I can make some test pages to show, for the time being I have added some screenshots!

kristinashu commented:
Thank you so much for this Daniel! Sorry I haven't gotten to this, I was engulfed by YouTube! The screen caps look great but yes, it will be helpful to see it in the review app.

sabrinang reviewed (changes requested):
I tested ideal & not ideal images ratios too. The 16:6 ratio works just fine but it gets a bit tall as it deviates
<img width="1228" alt="image" src="https://user-images.githubusercontent.com/2492510/125347856-4ca81d80-e310-11eb-802c-4729153685fa.png">

For images that are square or vertical at the various break points, it is getting quite tall at the large-small breakpoints. If staff upload images that aren't close to the ratio it would be helpful to have a max height closer to 400px for these breakpoints instead of 730px so the image isn't taking up most of the viewport.

sabrinang reviewed (commented):
It looked good from the zoom screen share but the review app for this is gone

Pomax commented:
@sabrinang the review apps get deleted after 5 days so we had to rebuild it - should be back up!

sabrinang reviewed (approved):
Thanks for setting up the review app again, image width radio buttons work well and looks fine on the page 👍

danielfmiranda commented:
Hi @Pomax thank you for the help getting the review app back up and running, I have made the requested changes/fixed the flake8 formatting issues/ made migrations and the PR is ready for re-review. Thanks again!

Pomax reviewed (changes requested):
Let's make sure the migration also copies over the "wide image" setting from before. This might be a little tricky, since blocks aren't actually Django models, so the uplift might be a little more involved than what you'd normally do for a field change in a model.

danielfmiranda commented on network-api/networkapi/mozfest/migrations/0016_add_image_width_select_field.py:
```diff
@@ -0,0 +1,63 @@
+# Generated by Django 3.0.14 on 2021-08-10 20:04
```
This file is made to add the new image_width radio select field.

danielfmiranda commented on network-api/networkapi/mozfest/migrations/0017_update_wide_images.py:
```diff
@@ -0,0 +1,65 @@
+# # Generated by Django 3.0.14 on 2021-08-10 18:19
```
This file loops through pages in the mozfest app, checks for any images that are set to "wide" and then updates the radio select field to be wide as well.

danielfmiranda commented on network-api/networkapi/mozfest/migrations/0018_remove_width_boolean.py:
```diff
@@ -0,0 +1,63 @@
+# Generated by Django 3.0.14 on 2021-08-10 20:07
```
This file then removes the original "wide_image" boolean field

danielfmiranda commented on network-api/networkapi/wagtailpages/migrations/0022_update_wide_images.py:
```diff
@@ -0,0 +1,72 @@
+# # Generated by Django 3.0.14 on 2021-08-10 18:19
```
This file loops through pages in the mofo app, checks for any images that are set to "wide" and then updates the radio select field to be wide as well.

danielfmiranda commented on network-api/networkapi/wagtailpages/migrations/0023_auto_20210810_2007.py:
```diff
@@ -0,0 +1,153 @@
+# Generated by Django 3.0.14 on 2021-08-10 20:07
```
This file then removes the original "wide_image" boolean field

danielfmiranda commented on network-api/networkapi/wagtailpages/migrations/0021_add_image_width_select_field.py:
```diff
@@ -0,0 +1,154 @@
+# Generated by Django 3.0.14 on 2021-08-10 20:04
```
This file is made to add the new image_width radio select field in the mofo app

Pomax commented on network-api/networkapi/mozfest/migrations/0017_update_wide_images.py:
```diff
@@ -0,0 +1,65 @@
+def update_annotated_image_blocks(qs):
+    # Loop through all pages
+    for page in qs:
+        print("\t", page)
```
rogue print statement? =)

danielfmiranda replied:
Removed! Marking as resolved

Pomax commented on network-api/networkapi/mozfest/migrations/0017_update_wide_images.py:
```diff
@@ -0,0 +1,65 @@
+def loop_through_pages_with_image_blocks(apps, schema):
+    # Only look through pages that ACTUALLY USE the Annotated Image Block streamfield.
+
+    pages_with_image_blocks = [
+        MozfestPrimaryPage.objects.all(),
+    ]
+
+    for page_qs in pages_with_image_blocks:
```
```suggestion
    for page in MozfestPrimaryPage.objects.all():
```

danielfmiranda replied:
Thanks for the catch pomax! I have added this to the latest rendition, going to mark this conversation as resolved!

Pomax commented on network-api/networkapi/mozfest/migrations/0017_update_wide_images.py:
```diff
@@ -0,0 +1,65 @@
+    ]
+
+    for page_qs in pages_with_image_blocks:
+        print("Pages:", page_qs.count())
```
rogue print?

danielfmiranda replied:
Removed! Marking as resolved

Pomax commented on network-api/networkapi/mozfest/migrations/0017_update_wide_images.py:
```diff
@@ -0,0 +1,65 @@
+
+    for page_qs in pages_with_image_blocks:
+        print("Pages:", page_qs.count())
+        update_annotated_image_blocks(page_qs)
```
Let's move that function's code here, since really the preamble is just a single line

danielfmiranda replied:
Thanks for the suggestion Pomax, I have now moved the functions code to this line instead of calling it like before. Marking this convo as resolved!

Pomax commented on network-api/networkapi/mozfest/migrations/0017_update_wide_images.py:
```diff
@@ -0,0 +1,65 @@
+        # Only loop through streamfield data if there's a `body` field on the page
+        if page.body:
+            # Looping through raw_data is a List of blocks.
+            for block in page.body.raw_data:
```
Note that this only works "for now". If some future version of Wagtail changes the property name, this migration will error out. Let's at the very least test for `hasattr(page.body, 'raw_data')` before we loop through it.

danielfmiranda replied:
Thank you both for taking the time to look at this PR and for the catch. I have added the check and am marking this conversation as resolved!

Pomax commented on network-api/networkapi/mozfest/migrations/0017_update_wide_images.py:
```diff
@@ -0,0 +1,65 @@
+                # If there is a type in the blocks, and the block type is a image,
+                # check if the image has the wide setting on
+                if 'type' in block and block['type'] == "image":
+                    print(block)
```
rogue print

danielfmiranda replied:
Removed! Marking as resolved

Pomax commented on network-api/networkapi/mozfest/migrations/0017_update_wide_images.py:
```diff
@@ -0,0 +1,65 @@
+                    print(block)
+                    # If so, set the new image_width value to wide
+                    if 'wide_image' in block['value'] and block['value']['wide_image'] == True:
+                        print("Found a wide image, updating.")
```
rogue print

danielfmiranda replied:
Removed! Marking as resolved

Pomax commented on network-api/networkapi/mozfest/migrations/0017_update_wide_images.py:
```diff
@@ -0,0 +1,65 @@
+                        block['value']['image_width'] = "wide"
+                        needs_saving = True
+
+            if needs_saving:
+                # If page is published already, continue to publish it.
+                # Otherwise just save a revision for draft history.
+                if page.live:
+                    print("\t\tPage is live. Publish it")
+                    page.save_revision().publish()
+                else:
+                    print("\t\tPage is draft.")
+                    page.save_revision()
```
```suggestion
            if needs_saving:
                page.save_revision()

                if page.live:
                    page.publish()
```

KalobTaulien replied:
`page` needs a revision to publish. 

```suggestion
            if needs_saving:
                revision = page.save_revision()
                
                if page.live:
                    revision.publish()
```

danielfmiranda replied:
Thanks for the catch everyone! I have made the update and am going to mark this as resolved.

Pomax reviewed (changes requested):
Remember not to put print statements in migrations: if you want to log progress, this should be a management command that we can run once, or it's a migration and should run silently. Though in this case, I'd say this should stay a migration (especially with the 0018 followup).

Pomax commented on network-api/networkapi/mozfest/migrations/0017_update_wide_images.py:
```diff
@@ -0,0 +1,65 @@
+def loop_through_pages_with_image_blocks(apps, schema):
+    # Only look through pages that ACTUALLY USE the Annotated Image Block streamfield.
+
+    pages_with_image_blocks = [
+        MozfestPrimaryPage.objects.all(),
+    ]
```
this creates a single-element list, with a queryset as first and only element, but then you take that queryset back out in the for loop, so that's several lines of code that ends up just being `MozfestPrimaryPage.objects.all()` again =)

danielfmiranda replied:
Hi @Pomax! Thanks for the catch, I have tried to implement the suggested change from above, where instead of creating a list,  we use `    for page in MozfestPrimaryPage.objects.all():`, however, I seem to be running into a interesting error when doing so. 

When I run `inv test-python` using the suggested change, the test suite then returns this error: 

`django.db.utils.ProgrammingError: column wagtailcore_page.title_en does not exist`
`LINE 1: ...re_page"."locale_id", "wagtailcore_page"."title", "wagtailco...`
`HINT:  Perhaps you meant to reference the column "wagtailcore_page.title".`

This error is returned even if I comment everything out underneath the "for" loop, and adding a print statement instead. 

However, if I add this line above the for loop like so: 
`    if MozfestPrimaryPage.objects.all().count() != 0:`
`       for page in MozfestPrimaryPage.objects.all():`

The test suite then runs OK, and the migration properly updates wide images when run. 

I have tried to troubleshoot for a while now but have not been able to find the reason behind this, however I do see that while running the test suite, if allowed to run all the way through, seems to run some SQL statements that add the "title_en" column to the test DB:

<img width="757" alt="image" src="https://user-images.githubusercontent.com/18314510/129284796-218e6f48-c837-4fba-97f6-26220bb98548.png">

Then after using [this WIP PR](https://github.com/mozilla/foundation.mozilla.org/blob/f92b8175ce710aa06d6e578791a0a35483853c79/network-api/networkapi/wagtailpages/migrations/0020_migrate_quote_blocks.py#L49) for reference regarding custom migrations, I noted that we were filtering for objects with the "locale_id" of 1, for English like so:

`    pages_with_quote_blocks = [`
       ` BlogPage.objects.filter(locale_id=1),`
       ` PrimaryPage.objects.filter(locale_id=1),`
      `  ModularPage.objects.filter(locale_id=1),`
      `  MozfestPrimaryPage.objects.filter(locale_id=1)`
  `  ]` 

I found that in the error that is being produced from the testing suite, it appears that the "locale_id" column **does** exist in the DB, so I tried to use the filter statement as well, but the same error persists. 

I was wondering if this has been something that you have run into before when making a custom migration? 

Thanks!

danielfmiranda replied:
Thanks for the detailed explanation Pomax! I have made the update, and am going to mark this conversation as resolved

Pomax commented:
Also note that `inv test-python` fails.

danielfmiranda commented:
Per request of Pomax, moving this to blocked until we finish the work with the wagtail-localise-2021 branch

Pomax commented:
localize landed 🎉

danielfmiranda commented:
Thanks for the update @Pomax! 

I have made the requested changes, added updated migrations,  and have re-requested review!

Pomax commented on network-api/networkapi/mozfest/migrations/0018_update_wide_images.py:
```diff
@@ -0,0 +1,48 @@
+
+def loop_through_pages_with_image_blocks(apps, schema):
+
+    for page in MozfestPrimaryPage.objects.all():
```
it looks like we'll need an extra check for page aliasses now that we use wagtail-localize:

```suggestion
    for page in MozfestPrimaryPage.objects.all():
        if hasattr(page, 'alias_of') and page.alias_of != None:
            # This is a page alias, rather than a real page
            continue
```

Otherwise we'll run into this problem when we try to apply it to our real data:

```
Running migrations:
  Applying mozfest.0017_auto_20210818_0231... OK
  Applying mozfest.0018_update_wide_images... OK
  Applying mozfest.0019_auto_20210818_0231... OK
  Applying wagtailpages.0031_auto_20210818_0231... OK
  Applying wagtailpages.0032_update_wide_images...Traceback (most recent call last):
  File "network-api/manage.py", line 34, in <module>
    execute_from_command_line(sys.argv)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/core/management/__init__.py", line 401, in execute_from_command_line
    utility.execute()
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/core/management/__init__.py", line 395, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/core/management/base.py", line 330, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/core/management/base.py", line 371, in execute
    output = self.handle(*args, **options)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/core/management/base.py", line 85, in wrapped
    res = handle_func(*args, **kwargs)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/core/management/commands/migrate.py", line 245, in handle
    fake_initial=fake_initial,
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/db/migrations/executor.py", line 117, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/db/migrations/executor.py", line 147, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/db/migrations/executor.py", line 227, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/db/migrations/migration.py", line 124, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/db/migrations/operations/special.py", line 190, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/app/network-api/networkapi/wagtailpages/migrations/0032_update_wide_images.py", line 53, in loop_through_pages_with_image_blocks
    update_annotated_image_blocks(page_qs)
  File "/app/network-api/networkapi/wagtailpages/migrations/0032_update_wide_images.py", line 38, in update_annotated_image_blocks
    revision = page.save_revision()
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/wagtail/core/models.py", line 1407, in save_revision
    "save_revision() was called on an alias page. "
RuntimeError: save_revision() was called on an alias page. Revisions are not required for alias pages as they are an exact copy of another page.
1
```

Pomax commented on network-api/networkapi/wagtailpages/migrations/0032_update_wide_images.py:
```diff
@@ -0,0 +1,63 @@
+
+def update_annotated_image_blocks(qs):
+    # Loop through all pages
+    for page in qs:
```
```suggestion
    for page in qs:
        if hasattr(page, 'alias_of') and page.alias_of != None:
            # This is a page alias, rather than a real page
            continue
```

Pomax reviewed (approved):
this seems to be working nicely
